cpp/custrings/category/NVCategory.cu

```diff
@@ -760,7 +760,10 @@ int NVCategory::get_value(unsigned int index)
     int* d_map = pImpl->getMapPtr();
     int rtn = -1;
     if( d_map )
+    {
         CUDA_TRY(cudaMemcpyAsync(&rtn,d_map+index,sizeof(int),cudaMemcpyDeviceToHost))
+        CUDA_TRY(cudaStreamSynchronize(0))
```

Comment:
This is effectively the same thing as a `cudaDeviceSynchronize()`.

Reply:
Oh right, because of the special semantics of stream 0.  @davidwendt wanted to avoid synchronizing the entire device to fix these races, but that's going to require the non-default stream to accomplish.

I see NVCategoryImpl has a `stream_id`, but I believe it's always 0.  Does it sound reasonable to change this to use `stream_id` for the transfer?  That way it would be async relative to other non-default streams if/when `stream_id` becomes useful.  These calls would be device-synchronizing until that time and for categories that remain on stream 0.

Another approach would be to create an ephemeral stream within these methods just for this transfer, but I was under the impression creating and destroying streams was a non-trivial cost.

Reply:
Do not create temporary streams to solve this. Did this race condition actually occur in a use case or are you just fixing this from inspection?  This code will be reworked to handle streams.  I suppose the least evil here is to use the `stream_id` for now.

Reply:
Since these memcpys are just using the default stream anyways, how about just switching them to normal `cudaMemcpy`s?

Reply:
I'm ok with that.

Reply:
> Did this race condition actually occur in a use case or are you just fixing this from inspection?

From inspection.

>  I suppose the least evil here is to use the stream_id for now.

Fixing `get_value` to use `pImpl->stream_id` rather than stream 0 is easy, but it's less clear how best to fix `get_value_bounds`.  If I just change the last copy at the end to use `stream_id` then potentially this method could start mixing stream usage between stream 0 and some non-zero stream which I believe is going to be another race condition to solve.  To avoid that future landmine I started to change the method to use `stream_id` throughout, but I ran into https://github.com/rapidsai/rmm/issues/120.

Because the stream thing is much more involved than just adding a stream specifier to a couple of places, I propose we fix the synchronization issue in `get_value_bounds` by changing the last copy to `cudaMemcpy`.  Until NVCategory supports the non-default stream, any stream synchronization is going to be a full device sync anyway, and I'd rather not hold up this bugfix on stream support being implemented.  Thoughts?

Reply:
Ah, sorry for the comment race.  I'll fix it up to use cudaMemcpy for now.